maggy/core/rpc.py

```diff
@@ -217,20 +217,120 @@ def _handle_message(self, sock, msg, exp_driver):
+            send["type"] = "QUERY"
+            send["data"] = self.reservations.done()
+        elif msg_type == "METRIC":
+            # add metric msg to the exp driver queue
+            exp_driver.add_message(msg)
+
+            if msg["trial_id"] is None:
+                send["type"] = "OK"
+                MessageSocket.send(self, sock, send)
+                return
+            elif msg["trial_id"] is not None:
+                if msg.get("data", None) is None:
+                    send["type"] = "OK"
+                    MessageSocket.send(self, sock, send)
+                    return
+
+            # lookup executor reservation to find assigned trial
+            trialId = msg["trial_id"]
+            # get early stopping flag, should be False for ablation
+            flag = exp_driver.get_trial(trialId).get_early_stop()
+
+            if flag:
+                send["type"] = "STOP"
+            else:
+                send["type"] = "OK"
+        elif msg_type == "FINAL":
+            # reset the reservation to avoid sending the same trial again
+            self.reservations.assign_trial(msg["partition_id"], None)
+
+            send["type"] = "OK"
+
+            # add metric msg to the exp driver queue
+            exp_driver.add_message(msg)
+        elif msg_type == "GET":
+            # lookup reservation to find assigned trial
+            trial_id = self.reservations.get_assigned_trial(msg["partition_id"])
+
+            # trial_id needs to be none because experiment_done can be true but
+            # the assigned trial might not be finalized yet
+            if exp_driver.experiment_done and trial_id is None:
+                send["type"] = "GSTOP"
+            else:
+                send["type"] = "TRIAL"
+
+            send["trial_id"] = trial_id
+
+            # retrieve trial information
+            if trial_id is not None:
+                send["data"] = exp_driver.get_trial(trial_id).params
+                exp_driver.get_trial(trial_id).status = Trial.RUNNING
+            else:
+                send["data"] = None
+        elif msg_type == "LOG":
+            # get data from experiment driver
+            result, log = exp_driver._get_logs()
+
+            send["type"] = "OK"
+            if log:
+                send["ex_logs"] = log
+            else:
+                send["ex_logs"] = None
+            send["num_trials"] = exp_driver.num_trials
+            send["to_date"] = result["num_trials"]
+            send["stopped"] = result["early_stopped"]
+            send["metric"] = result["best_val"]
+        else:
+            send["type"] = "ERR"
+
+        MessageSocket.send(self, sock, send)
+
+    def get_assigned_trial_id(self, partition_id):
+        """Returns the id of the assigned trial, given a ``partition_id``.
+
+        Arguments:
+            partition_id {[type]} -- [description]
+
+        Returns:
+            trial_id
+        """
+        return self.reservations.get_assigned_trial(partition_id)
+
```

Comment:
This is part of the old RPC logic. Seems to be a merge error. Has to be removed.